main/src/util/profiler.hpp

```diff
@@ -0,0 +1,147 @@
+#pragma once
+
+#include <vector>
+#include <fstream>
+#include <mpi.h>
+
+namespace sphexa
+{
+
+class Profiler
+{
+private:
+    std::vector<float*> timeSteps;     // vector of timesteps for all ranks
+    std::vector<float>  meanPerStep;   // mean (average)
+    std::vector<double> stdevPerStep;  // standard deviation
+    std::vector<double> covPerStep;    // c.o.v = std / mean
+    std::vector<double> lambdaPerStep; // lambda = (Lmax / Lmean - 1) * 100%
+    std::vector<float*> vectorMetric;  // vector based metric calculation
+    std::vector<double> I_2PerStep;    // distance to zero for vector based metric, name tentative
+    std::vector<float>  g1PerStep;     // skewness
+    std::vector<float>  g2PerStep;     // kurtosis
+    std::vector<float>  totalTimeStep; // total time-step durations
+    std::vector<size_t> numLocalParts; // number of local particles
+    std::vector<float>  funcTimes;     // vector of timesteps for all ranks
+    int                 _numRanks;
+    int                 _rank;
+    std::ofstream       profilingFile;
+
+public:
+    Profiler(int rank)
+        : _rank(rank)
+    {
+        MPI_Comm_size(MPI_COMM_WORLD, &_numRanks);
+        std::string filename = "profiling-" + std::to_string(_numRanks) + "Ranks.txt";
+        if (_rank == 0) profilingFile.open(filename);
+    }
+    ~Profiler() {}
+
+    void printMetrics(int iteration)
+    {
+        profilingFile << meanPerStep.at(iteration) << "," << stdevPerStep.at(iteration) << ","
+                      << covPerStep.at(iteration) << "," << I_2PerStep.at(iteration) << ","
+                      << lambdaPerStep.at(iteration);
+        for (int i = 0; i < _numRanks; i++)
+        {
+            profilingFile << "," << vectorMetric.at(iteration)[i];
+        }
+    }
+
+    void printProfilingInfo()
+    {
```

Comment:
I would only save the raw data, i.e. `funcTimes` plus relevant meta data, e.g. how many ranks, how many measuring points per step and their name string labels. Which means it would make the analysis more robust and easier to deal with and implement if you saved the timing data as HDF5 file attributes.